Request certificate API (#29310)
fleetdm/fleet: docs/REST API/rest-api.md

```
---

## Certificates

> 

Author: Noah: Maybe "List certificate authorities" here.

> - [Generate certificate](#request-certificate)

Author: ```suggestion
- [Request certificate](#request-certificate)
```

> ### Generate certificate

Author: ```suggestion
### Request certificate
```

> Generates a base64 encoded certificate (`.pem`) from the specified certificate authority (CA).

Author: ```suggestion
Requests a base64 encoded certificate (`.pem`) from the specified certificate authority (CA).
```

> `POST /api/v1/fleet/certificates/generate`

Author: ```suggestion
`POST /api/v1/fleet/certificates`
```

#### Parameters

| Name     | Type    | In   | Description                                 |
| -------- | ------- | ---- | ------------------------------------------- |

> | certificate_authority_name   | string | body | **Required.** The certificate authority (CA) name in Fleet. See your CAs [here](). |

Author: `id` would be more consistent with the rest of the Fleet API

> | certificate_authority_name   | string | body | **Required.** The certificate authority (CA) name in Fleet. See your CAs [here](). |

Author: ```suggestion
| certificate_authority_is   | string | body | **Required.** The certificate authority (CA) name in Fleet. See your CAs [here](). |
```

> | certificate_signing_request       | string | body | **Required.** The signed certificate signing request (CSR). Noah: Does NDES and Custom SCEP need a CSR?        |

Author: Noah: Maybe we go with shorthand `csr` here?

> | common_name   | integer | body | Certificates will have this CN in the subject. |

Author: ```suggestion
| common_name   | string | body | Certificates will have this CN in the subject. |
```

> | certificate_signing_request       | string | body | **Required.** The signed certificate signing request (CSR). Noah: Does NDES and Custom SCEP need a CSR?        |
> | common_name   | integer | body | Certificates will have this CN in the subject. |
> | user_principal_name   | string | body | Certificates will have this UPN. |
> | subject_alternative_name   | string | body | Certificates will have this SAN. |
> | uniform_resource_identifier  | string | body | Certificates will have this URI. |

Author: ```suggestion
| csr       | string | body | **Required.** The signed certificate signing request (CSR). Noah: Does NDES and Custom SCEP need a CSR?        |
| cn   | integer | body | Certificates will have this common name (CN) in the subject. |
| upn   | string | body | Certificates will have this user principal name (UPN). |
| san   | string | body | Certificates will have this subject alternative name (SAN). |
| uri  | string | body | Certificates will have this uniform resource identifier (URI). |
```

| idp_token | string | body | Active session token from your identity provider you configured for [end user authentication](#mdm-end-user-authentication). Required if you're enforcing end user authentication |

#### Example

> `POST /api/v1/fleet/certificates/generate`

Author: ```suggestion
`POST /api/v1/fleet/certificates`
```

##### Request

```json
{
  "certificate_authority_name": "WIFI_CERTIFICATES",

>   "csr": "-----BEGIN CERTIFICATE REQUEST-----\nMIIC/jCCAeYCAQAwITEfMB0GA1UEAwwWQ2lzY29Vc2VyTmV0d29ya0FjY2VzczCC\nASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBALJZtbxathh+RfK+Z613ar4E\nYSIem8yAvv2JZJtopjD3noy1yF+nGRyF/ocm+FhYvjR5u7teJXlcv24tAAHuWL4U\nuPIql0Slakjdsfl098salkj324lkjmtElWDi6XRjUIXEj1zyCnZTCxGmyHcYB/+f3fyv/\ngZ8SkPqocNOCpX6cSW8hxOlaF9aZUC+xMHRdjQgxQ79hleb5K/n2gCJjiW1sV0Es\nRg+MX0cbPCpahpzlvIAkzA7TTUTOd7ZN+V0GW0fH86uMstrqeW2QUuZmSDC9fNyj\nQhk6n5iURaHXdFjSmyrhW5AVvw1nIblHodhUtD6J+g9kjhBg1frss3ndQtnNrnMC\nAwEAAaCBlzCkldflkjc098dlkj2KoZIhvcNAQkOMYGGMIGDMIGABgNVHREEeTB3ggljaXNjby5j\nb22BEWthYW53YXJAY2lzY28uY29thjRJRDpGbGVldERNOkdVSUQ6Y2FkMTM4OTEt\nMzU3Ni00NzhmLTk1MzAtZmM1Y2VlZTEzZTkwoCEGCisGAQQBgjcUAgOgEwwRa2Fh\nbndhckBjaXNjby5jb20wDQYJKoZIhvcNAQELBQADggEBAH2U6Or14b4O22YjM22k\nXI9QDC5P+sDczcLjivv4MyXQL1ks8R6B1nXCrOmiLPPLaZ09f+UkeMnyuGAxW8Ce\n6LTKquwvlifZ+5TjyANz0I/d9ETLQF2MTphEZd4ySNLtq2RwYyDOBKaxMdW0sUsd\n6M3WyAuTBVgBkTVIqbMJBzFsgXSrr2a0LJEHszOO2BN3yT5muDQsKPJ1uXL7tNUv\n16pGaYpQZR8yGAmWyISHhAyLaJ1N1R8L77SLxdd/Sj7RunNNxqFqaEgIJMgsyu08\nGharLkQcIoW7qPHZuaLa54xMF/s/vfKH6rgGbbCAgw9kw8Klt+6H3OH1FSMeRfZ/\nDWs=\n-----END CERTIFICATE REQUEST-----",

Author: ```suggestion
  "csr": "-----BEGIN CERTIFICATE REQUEST-----MIIC/jCCAeYCAQAwITEfMB0GA1UEAwwWQ2lzY29Vc2VyTmV0d29ya0FjY2VzczCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBALJZtbxathh+RfK+Z613ar4EYSIem8yAvv2JZJtopjD3noy1yF+nGRyF/ocm+FhYvjR5u7teJXlcv24tAAHuWL4UuPIql0SotR+VBMLH04Qs97nsmtElWDi6XRjUIXEj1zyCnZTCxGmyHcYB/+f3fyv/gZ8SkPqocNOCpX6cSW8hxOlaF9aZUC+xMHRdjQgxQ79hleb5K/n2gCJjiW1sV0EsRg+MX0cbPCpahpzlvIAkzA7TTUTOd7ZN+V0GW0fH86uMalksjdf;klja;sdp-0asdopifjlk;j234;lk;lkjalscvncm,.vkasuda0s8d-fasd0[fadsflakdksl;fjas;lkdfjalsdjf;lkasdjflasjdklasdlf;oasd8979a80s7d08f9a7sdfstrqeW2QUuZmSDC9fNyjQhk6n5iURaHXdFjSmyrhW5AVvw1nIblHodhUtD6J+g9kjhBg1frss3ndQtnNrnMCAwEAAaCBlzCBlAYJKoZIhvcNAQkOMYGGMIGDMIGABgNVHREEeTB3ggljaXNjby5jb22BEWthYW53YXJAY2lzY28uY29thjRJRDpGbGVldERNOkdVSUQ6Y2FkMTM4OTEtMzU3Ni00NzhmLTk1MzAtZmM1Y2VlZTEzZTkwoCEGCisGAQQBgjcUAgOgEwwRa2FhbndhckBjaXNjby5jb20wDQYJKoZIhvcNAQELBQADggEBAH2U6Or14b4O22YjM22kXI9QDC5P+sDczcLjivv4MyXQL1ks8R6B1nXCrOmiLPPLaZ09f+UkeMnyuGAxW8Ce6LTKquwvlifZ+5TjyANz0I/d9ETLQF2MTphEZd4ySNLtq2RwYyDOBKaxMdW0sUsd6M3WyAuTBVgBkTVIqbMJBzFsgXSrr2a0LJEHszOO2BN3yT5muDQsKPJ1uXL7tNUv16pGaYpQZR8yGAmWyISHhAyLaJ1N1R8L77SLxdd/Sj7RunNNxqFqaEgIJMgsyu08GharLkQcIoW7qPHZuaLa54xMF/s/vfKH6rgGbbCAgw9kw8Klt+6H3OH1FSMeRfZ/DWs=-----END CERTIFICATE REQUEST-----",
```

>   "certificate_authority_name": "WIFI_CERTIFICATES",
>   "certificate_signing_request": "-----BEGIN CERTIFICATE REQUEST-----\nMIIC/jCCAeYCAQAwITEfMB0GA1UEAwwWQ2lzY29Vc2VyTmV0d29ya0FjY2VzczCC\nASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBALJZtbxathh+RfK+Z613ar4E\nYSIem8yAvv2JZJtopjD3noy1yF+nGRyF/ocm+FhYvjR5u7teJXlcv24tAAHuWL4U\nuPIql0Slakjdsfl098salkj324lkjmtElWDi6XRjUIXEj1zyCnZTCxGmyHcYB/+f3fyv/\ngZ8SkPqocNOCpX6cSW8hxOlaF9aZUC+xMHRdjQgxQ79hleb5K/n2gCJjiW1sV0Es\nRg+MX0cbPCpahpzlvIAkzA7TTUTOd7ZN+V0GW0fH86uMstrqeW2QUuZmSDC9fNyj\nQhk6n5iURaHXdFjSmyrhW5AVvw1nIblHodhUtD6J+g9kjhBg1frss3ndQtnNrnMC\nAwEAAaCBlzCkldflkjc098dlkj2KoZIhvcNAQkOMYGGMIGDMIGABgNVHREEeTB3ggljaXNjby5j\nb22BEWthYW53YXJAY2lzY28uY29thjRJRDpGbGVldERNOkdVSUQ6Y2FkMTM4OTEt\nMzU3Ni00NzhmLTk1MzAtZmM1Y2VlZTEzZTkwoCEGCisGAQQBgjcUAgOgEwwRa2Fh\nbndhckBjaXNjby5jb20wDQYJKoZIhvcNAQELBQADggEBAH2U6Or14b4O22YjM22k\nXI9QDC5P+sDczcLjivv4MyXQL1ks8R6B1nXCrOmiLPPLaZ09f+UkeMnyuGAxW8Ce\n6LTKquwvlifZ+5TjyANz0I/d9ETLQF2MTphEZd4ySNLtq2RwYyDOBKaxMdW0sUsd\n6M3WyAuTBVgBkTVIqbMJBzFsgXSrr2a0LJEHszOO2BN3yT5muDQsKPJ1uXL7tNUv\n16pGaYpQZR8yGAmWyISHhAyLaJ1N1R8L77SLxdd/Sj7RunNNxqFqaEgIJMgsyu08\nGharLkQcIoW7qPHZuaLa54xMF/s/vfKH6rgGbbCAgw9kw8Klt+6H3OH1FSMeRfZ/\nDWs=\n-----END CERTIFICATE REQUEST-----",
>   "common_name": "NetworkAccess",
>   "user_principal_name": "username@example.com",
>   "subject_alternative_name": "example.com",
>   "uniform_resource_identifier": "username@example.com",

Author: ```suggestion
  "ca_id": "WIFI_CERTIFICATES",
  "csr": "-----BEGIN CERTIFICATE REQUEST-----\nMIIC/jCCAeYCAQAwITEfMB0GA1UEAwwWQ2lzY29Vc2VyTmV0d29ya0FjY2VzczCC\nASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBALJZtbxathh+RfK+Z613ar4E\nYSIem8yAvv2JZJtopjD3noy1yF+nGRyF/ocm+FhYvjR5u7teJXlcv24tAAHuWL4U\nuPIql0Slakjdsfl098salkj324lkjmtElWDi6XRjUIXEj1zyCnZTCxGmyHcYB/+f3fyv/\ngZ8SkPqocNOCpX6cSW8hxOlaF9aZUC+xMHRdjQgxQ79hleb5K/n2gCJjiW1sV0Es\nRg+MX0cbPCpahpzlvIAkzA7TTUTOd7ZN+V0GW0fH86uMstrqeW2QUuZmSDC9fNyj\nQhk6n5iURaHXdFjSmyrhW5AVvw1nIblHodhUtD6J+g9kjhBg1frss3ndQtnNrnMC\nAwEAAaCBlzCkldflkjc098dlkj2KoZIhvcNAQkOMYGGMIGDMIGABgNVHREEeTB3ggljaXNjby5j\nb22BEWthYW53YXJAY2lzY28uY29thjRJRDpGbGVldERNOkdVSUQ6Y2FkMTM4OTEt\nMzU3Ni00NzhmLTk1MzAtZmM1Y2VlZTEzZTkwoCEGCisGAQQBgjcUAgOgEwwRa2Fh\nbndhckBjaXNjby5jb20wDQYJKoZIhvcNAQELBQADggEBAH2U6Or14b4O22YjM22k\nXI9QDC5P+sDczcLjivv4MyXQL1ks8R6B1nXCrOmiLPPLaZ09f+UkeMnyuGAxW8Ce\n6LTKquwvlifZ+5TjyANz0I/d9ETLQF2MTphEZd4ySNLtq2RwYyDOBKaxMdW0sUsd\n6M3WyAuTBVgBkTVIqbMJBzFsgXSrr2a0LJEHszOO2BN3yT5muDQsKPJ1uXL7tNUv\n16pGaYpQZR8yGAmWyISHhAyLaJ1N1R8L77SLxdd/Sj7RunNNxqFqaEgIJMgsyu08\nGharLkQcIoW7qPHZuaLa54xMF/s/vfKH6rgGbbCAgw9kw8Klt+6H3OH1FSMeRfZ/\nDWs=\n-----END CERTIFICATE REQUEST-----",
  "cn": "NetworkAccess",
  "upn": "username@example.com",
  "san": "example.com",
  "uri": "username@example.com",
```

```
---

## Certificates

- [List certificate authorities (CAs)](#list-certificate-authorities-cas)
- [Request certificate](#request-certificate)

### List certificate authorities (CAs)

TODO

### Request certificate

Requests a base64 encoded certificate (`.pem`) from the specified certificate authority (CA).

`POST /api/v1/fleet/certificates`

#### Parameters

| Name     | Type    | In   | Description                                 |
| -------- | ------- | ---- | ------------------------------------------- |
| ca_id   | string | body | **Required.** The certificate authority (CA) name in Fleet. See your CAs [here](#list-certificate-authorities-cas). |

> | csr       | string | body | **Required.** The signed certificate signing request (CSR). Noah: Does NDES and Custom SCEP need a CSR?        |

Author: ```suggestion
| csr       | string | body | The signed certificate signing request (CSR). If left unspecified, Fleet will create the CSR with a shared key.        |
```

> | csr       | string | body | The signed certificate signing request (CSR). If left unspecified, Fleet will create the CSR with a shared key.        |
> | cn   | integer | body | Certificates will have this common name (CN) in the subject. |
> | upn   | string | body | Certificates will have this user principal name (UPN). |
> | san   | string | body | Certificates will have this subject alternative name (SAN). |
> | uri  | string | body | Certificates will have this uniform resource identifier (URI). |

Author: ```suggestion
| csr       | string | body | The signed certificate signing request (CSR). If left unspecified, Fleet will create the CSR with a shared key.        |
| cn   | integer | body | Certificates will have this common name (CN) in the subject. If left unspecified, Fleet will use the CN in the template configured in the CA. |
| upn   | string | body | Certificates will have this user principal name (UPN). If left unspecified, Fleet will use the CN in the template configured in the CA. |
| san   | string | body | Certificates will have this subject alternative name (SAN). If left unspecified, Fleet will use the CN in the template configured in the CA. |
| uri  | string | body | Certificates will have this uniform resource identifier (URI). If left unspecified, Fleet will use the CN in the template configured in the CA. |
```

> | uri  | string | body | Certificates will have this uniform resource identifier (URI). If left unspecified, Fleet will use the CN in the template configured in the CA. |

Author: ```suggestion
| csr       | string | body | The signed certificate signing request (CSR). If left unspecified, Fleet will create the CSR with a shared key.        |
| cn   | integer | body | Certificates will have this common name (CN) in the subject. If left unspecified, Fleet will use the CN in the template configured in the CA. |
| upn   | string | body | Certificates will have this user principal name (UPN). If left unspecified, Fleet will use the UPN from the template you set up in the CA. |
| san   | string | body | Certificates will have this subject alternative name (SAN). If left unspecified, Fleet will use the SAN from the template you set up in the CA. |
| uri  | string | body | Certificates will have this uniform resource identifier (URI). If left unspecified, Fleet will use the URI from the template you set up in the CA. |
```
